tensorflow_probability/python/distributions/two_piece_student_t.py

```diff
@@ -0,0 +1,847 @@
+
+  return (
+      samples,
+      broadcast_ddf * partial_df + broadcast_dskewness * partial_skewness)
+
+
+@tfp_custom_gradient.custom_gradient(
+    vjp_fwd=_two_piece_student_t_sample_fwd,
+    vjp_bwd=_two_piece_student_t_sample_bwd,
+    jvp_fn=_two_piece_student_t_sample_jvp,
+    nondiff_argnums=(0,))
+def _two_piece_student_t_sample_with_gradient(sample_shape, df, skewness, seed):
+  """Generate samples from Two-Piece Student's t-distribution.
+
+  The distribution is the Two-Piece Student's t-distribution with degree of
+  freedom `df`, location zero, scale one, and skewness `skewness`. To change
+  the location and scale, use:
+
+  ```none
+  loc + scale * samples
+  ```
+
+  The samples are pathwise differentiable using the approach of [1].
+
+  Args:
+    sample_shape: 0D or 1D `int32` `Tensor`. Shape of the generated samples.
+    df: Floating-point tensor; the degree(s) of freedom of the
+      distribution(s). Must contain only positive values.
+    skewness: Floating-point tensor; the skewness(es) of the distribution(s).
+    seed: PRNG seed; see `tfp.random.sanitize_seed` for details.
+
+  Returns:
+    A tensor with shape `sample_shape`.
+
+  References:
+    [1]: Michael Figurnov, Shakir Mohamed, and Andriy Mnih.
+         Implicit Reparameterization Gradients. In _Advances in Neural
+         Information Processing Systems_, 31, 2018.
+         https://arxiv.org/abs/1805.08498
+  """
+  return _two_piece_student_t_sample_no_gradient(
+      sample_shape, df, skewness, seed)
+
+
+def random_two_piece_student_t(sample_shape, df, skewness, seed=None):
+  """Generate samples from Two-Piece Student's t-distribution.
+
+  The distribution is the Two-Piece Student's t-distribution with degree of
+  freedom `df`, location zero, scale one, and skewness `skewness`. To change
+  the location and scale, use:
+
+  ```none
+  loc + scale * samples
+  ```
+
+  The samples are pathwise differentiable using the approach of [1].
+
+  Note that `skewness` can be negative.
+
+  Args:
+    sample_shape: 0D or 1D `int32` `Tensor`. Shape of the generated samples.
+    df: Floating-point tensor; the degree(s) of freedom of the
+      distribution(s). Must contain only positive values.
+    skewness: Floating-point tensor; the skewness(es) of the distribution(s).
+    seed: PRNG seed; see `tfp.random.sanitize_seed` for details.
+
+  Returns:
+    A tensor with shape `sample_shape`.
+
+  References:
+    [1]: Michael Figurnov, Shakir Mohamed, and Andriy Mnih.
+         Implicit Reparameterization Gradients. In _Advances in Neural
+         Information Processing Systems_, 31, 2018.
+         https://arxiv.org/abs/1805.08498
+  """
+  sample_shape = ps.convert_to_shape_tensor(sample_shape, dtype_hint=tf.int32)
+  df = tf.convert_to_tensor(df)
+  skewness = tf.convert_to_tensor(skewness)
+  seed = samplers.sanitize_seed(seed, salt='two_piece_student_t')
+
+  return _two_piece_student_t_sample_with_gradient(
```

Comment:
This came up in the internal review: could we instead sample from a two-piece normal and divide by a gamma sample? The motivation is to avoid the custom gradient implementation.

Reply:
Hi @SiegeLordEx,

Excellent question! And it seems you're right. In the following notebooks, I implemented this new sampler and compared the theoretical and empirical means, standard deviations, and densities:
- [Right-Skewed Two-Piece Student's t-distribution](https://gist.github.com/leandrolcampos/450e664e365f96486bc815f62d310ed4)
- [Left-Skewed Two-Piece Student's t-distribution](https://gist.github.com/leandrolcampos/02293c9b3e0054fcda658026e78f8b15)
- [Symmetric Two-Piece Student's t-distribution](https://gist.github.com/leandrolcampos/1f815119d87bbd1b673311af19bdf40a)

For now, I have no idea about which strategy to use to mathematically prove the equivalence between the two samplers. Some advice? And would such proof be necessary?

Intuitively, this equivalence makes sense. Indeed, the Two-Piece Student's t-distribution results from joining, at the location, the corresponding halves of two Student's t-distributions with same degree of freedom and location but different scales. Each of these Student's t-distributions corresponds to a Normal distribution that can be used to generate samples from the former. The two corresponding Normal distributions would create the Two-Piece Normal distribution that we can use to generate samples from the original Two-Piece Student's t-distribution.

I inspected four R packages that contain samplers from this distribution (or from one of its equivalent alternative parameterizations): [gamlss.dist](https://rdrr.io/cran/gamlss.dist/src/R/ST3.R#sym-rST3), [skewt](https://rdrr.io/cran/skewt/src/R/skewt.R#sym-rskt), [fGarch](https://github.com/cran/fGarch/blob/7d85dd50aab0ed4654f7b39c4b810f75df4f684a/R/dist-sstd.R#L298), and [rugarch](https://github.com/cran/rugarch/blob/c9474fc9defb3854233230b55f6a9b8f3a860e3a/src/distributions.c#L420). None of them implement the idea of this new sampler. The first two packages generate samples using the quantile function. The other two generate samples using the Student's t-distribution sampler.

**Should I update this PR, implementing this new sampler?**

Reply:
I don't think a proof is necessary, you can see it's right by inspecting the implementation of two piece normal and two piece student t. And yeah, I think if you could implement it that'd be good, and simplify the code significantly since we won't need to maintain the custom gradient.